dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java

```diff
@@ -91,13 +91,25 @@
 
         // TODO(dfox): P2C optimization when we have high number of nodes to save CPU?
         // http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf
-        return preShuffled.stream()
-                .map(MutableChannelWithStats::computeScore)
-                .sorted(Comparator.comparingInt(SortableChannel::getScore))
-                .map(channel -> channel.delegate.maybeExecute(endpoint, request))
-                .filter(Optional::isPresent)
-                .map(Optional::get)
-                .findFirst();
+        List<SortableChannel> sortedByScore = sortByScore(preShuffled);
+
+        for (SortableChannel channel : sortedByScore) {
+            Optional<ListenableFuture<Response>> maybe = channel.delegate.maybeExecute(endpoint, request);
+            if (maybe.isPresent()) {
+                return maybe;
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private static List<SortableChannel> sortByScore(List<MutableChannelWithStats> preShuffled) {
+        List<SortableChannel> sortedByScore = new ArrayList<>();
+        for (MutableChannelWithStats channel : preShuffled) {
+            sortedByScore.add(channel.computeScore());
+        }
+        sortedByScore.sort(Comparator.comparingInt(SortableChannel::getScore));
```

Comment:
We can extract the comparator to a static field

Reply:
Or maybe even just implement comparable?

Reply:
Yep, the two should be equivalent from a performance point of view. I slightly prefer comparators over implementing compareTo, but no strong opinion.